File: EnvelopeASP/Controllers/AccountController.cs
using EnvelopeASP.Models;
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Authentication.Cookies;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace EnvelopeASP.Controllers
{
    public class AccountController : Controller
    {

        public AccountController()
        {
        }

        public IActionResult Index()
        {
            return View();
        }

        public async Task<IActionResult> SignUp(SignUp model)
        {
            if(Request.Method== "POST" && ModelState.IsValid)
            {
                model.Email = model.Email.Trim();
                if(model.Password.Length > 1024)
                {
                    ViewBag.Error = "The password is a bit big, rejected";
                    return View(model);
                }
                if(model.Password != model.ConfirmPassword)
                {
                    ViewBag.Error = "Passwords do not match";
                    return View(model);
                }
                var passwordConfig = new PasswordConfig();
                var hash = passwordConfig.GenerateHash(model.Password);
                await Procedures.InsertUser(model.Email, hash, passwordConfig);

                return Redirect("Login");
            } else
            {
                model ??= new SignUp();
                return View(model);
            }
        }

        public async Task<IActionResult> Login(Login model, string? ReturnUrl)
        {
            if (string.IsNullOrWhiteSpace(ReturnUrl))
            {
                ReturnUrl = "/";
            }
            if (User.Identity is not null && User.Identity.IsAuthenticated)
            {
                return Redirect(ReturnUrl);
            }
            if (Request.Method == "POST" && ModelState.IsValid)
            {
                model.Email = model.Email.Trim();
                
                var dbUserN = await Procedures.Sel_UserByEmail(model.Email);

                if (dbUserN != null)
                {
                    var isValid = DateTime.UtcNow > dbUserN.LockoutExpiry && dbUserN.PasswordConfig.Verify(model.Password, dbUserN.PasswordHash);

                    await Procedures.Upd_User_Login(dbUserN.Id, isValid);

                    if (isValid)
                    {
                        var claims = new Claim[2];
                        claims[0] = new Claim(ClaimTypes.NameIdentifier, dbUserN.Id.ToString());
                        claims[1] = new Claim(ClaimTypes.Name, model.Email);

                        var claimsIdentity = new ClaimsIdentity(claims, CookieAuthenticationDefaults.AuthenticationScheme);

                        var principle = new ClaimsPrincipal(claimsIdentity);

                        await HttpContext.SignInAsync(principle);

                        return Redirect(ReturnUrl);
                    }
                    
                }

                return View(model);
            }
            else
            {
                model ??= new Login();
                return View(model);
            }
        }

        public async Task<IActionResult> Logout()
        {
            if (User.Identity is not null && User.Identity.IsAuthenticated)
            {
                await HttpContext.SignOutAsync();
            }
            return Redirect("/");
        }

        [Authorize]
        public async Task<IActionResult> PasswordReset(SignUp model)
        {
            if (Request.Method == "POST" && ModelState.IsValid)
            {
                var oldPassword = model.Email.Trim();

                if (model.Password.Length > 1024)
                {
                    ViewBag.Error = "The new password is a bit big, rejected";
                    return View(model);
                }
                if (model.Password != model.ConfirmPassword)
                {
                    ViewBag.Error = "The new passwords do not match";
                    return View(model);
                }

                // Check Old Password
                var userNull = await Procedures.Sel_UserByEmail( User.Claims.First(f => f.Type == ClaimTypes.Name).Value);
                User user;
                if(userNull is null)
                {
                    ViewBag.Error = "Strange account not found error occured";
                    return View(model);
                }
                user = userNull;

                var isValid = DateTime.UtcNow > user.LockoutExpiry && user.PasswordConfig.Verify(oldPassword, user.PasswordHash);

                await Procedures.Upd_User_Login(user.Id, isValid);

                if(user.LockoutExpiry > DateTime.UtcNow)
                {
                    // password change is getting kinda sus
                    await HttpContext.SignOutAsync();
                    return Redirect("/");
                }

                if(isValid)
                {
                    // do on new password
                    var passwordConfig = new PasswordConfig();
                    var hash = passwordConfig.GenerateHash(model.Password);
                    var rowsEffected = await Procedures.UpdateUserPassword(Utils.GetUserIDFromClaims(User), hash, passwordConfig);
                } else
                {
                    ViewBag.Error = "Current password is invalid";
                    return View(model);
                }

                return Redirect("/");
            }
            else
            {
                model ??= new SignUp();
                return View(model);
            }
        }
    }
}

File: EnvelopeWeb/Controllers/AccountController.fs
namespace EnvelopeWeb.Controllers


open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Logging
open Models
open System.Threading.Tasks
open System.Security.Claims
open Microsoft.AspNetCore.Authentication.Cookies
open Microsoft.AspNetCore.Authentication
open Microsoft.AspNetCore.Authorization


type AccountController (logger : ILogger<AccountController>, dbConnectionGetter : DbConnection.IDbConnectionGetter) =
    inherit Controller()

    member this.Index () =
        this.View()

    member this.SignUp (model:Models.SignUp) : Task<IActionResult> =
        if this.Request.Method = "POST" && this.ModelState.IsValid then
            let viewBag = this.ViewData
            if model.Password.Length > 1024 then
                viewBag["Error"] <- "The password is a bit big, rejected"
                task { return this.View(model) }
            else if model.Password <> model.ConfirmPassword then
                viewBag["Error"] <- "Passwords do not match"
                task { return this.View(model) }
            else
                
                let passwordConfig = Password.makeDefaultPasswordConfig()
                let hash = Password.generateHash passwordConfig model.Password
                let email = model.Email.Trim()
                async {

                    let! t = Procedures.InsertUser dbConnectionGetter email hash passwordConfig

                    return this.Redirect("Login") :> IActionResult
                } |> Async.StartAsTask

        else
            task { return this.View(model) }


    member this.Login (model: Login) (returnUrl : string) : Task<IActionResult> =
        let returnUrl = if System.String.IsNullOrEmpty(returnUrl) then "/" else returnUrl
        if this.User.Identity.IsAuthenticated then
            task { return this.Redirect(returnUrl) }
        else if this.Request.Method = "POST" && this.ModelState.IsValid then
            let email = model.Email.Trim()
            async {
                let! dbUserN = Procedures.Sel_UserByEmail dbConnectionGetter email

                let returnResult : Async<IActionResult> =
                    match dbUserN with
                    | None -> async { return this.View(model) }
                    | Some dbUser ->
                            let isValid = System.DateTime.UtcNow > dbUser.LockoutExpiry && (Password.verify dbUser.PasswordConfig model.Password dbUser.PasswordHash)
                            async {
                                let! r = Procedures.Upd_User_Login dbConnectionGetter dbUser.Id isValid

                                if isValid then 
                                    let idClaim = new Claim(ClaimTypes.NameIdentifier, dbUser.Id.ToString())
                                    let nameClaim = new Claim(ClaimTypes.Name, email)
                                    let claims = [| idClaim; nameClaim; |]
                                    let claimsIdentity = new ClaimsIdentity(claims, CookieAuthenticationDefaults.AuthenticationScheme)
                                    let principle = new ClaimsPrincipal(claimsIdentity)

                                    let! r = this.HttpContext.SignInAsync(principle) |> Async.AwaitTask

                                    return this.Redirect(returnUrl)
                                else
                                    return this.View(model)
                            }
                            

                return! returnResult
                
            } |> Async.StartAsTask
        else
            task { return this.View(model) }

    member this.Logout() =
        async {
            if this.User.Identity.IsAuthenticated then
                do! this.HttpContext.SignOutAsync() |> Async.AwaitTask

            return this.Redirect("/")
        } |> Async.StartAsTask

    [<Authorize>]
    member this.PasswordReset (model:SignUp) : Task<IActionResult> =
        if this.Request.Method = "POST" && this.ModelState.IsValid then
            let viewBag = this.ViewData
           
            if model.Password.Length > 1024 then
                viewBag["Error"] <- "The new password is a bit big, rejected"
                task { return this.View(model) }
            else if model.Password <> model.ConfirmPassword then
                viewBag["Error"] <- "The new passwords do not match"
                task { return this.View(model) }
            else
                // check old password
                async {
                    let emailClaim = this.User.Claims |> Seq.find (fun f -> f.Type = ClaimTypes.Name)
                    let! optUser = Procedures.Sel_UserByEmail dbConnectionGetter emailClaim.Value

                    match optUser with
                    | None -> 
                        viewBag["Error"] <- "Strange user not found error occured"
                        return this.View(model) :> IActionResult
                    | Some dbUser ->
                        let isValid = System.DateTime.UtcNow > dbUser.LockoutExpiry && (Password.verify dbUser.PasswordConfig model.Password dbUser.PasswordHash)

                        return! async {
                            let! r = Procedures.Upd_User_Login dbConnectionGetter dbUser.Id isValid

                            if isValid then 
                                let passwordConfig = Password.makeDefaultPasswordConfig()
                                let hash = Password.generateHash passwordConfig model.Password
                                let uid = Utils.getUserIDFromClaims this.User
                                let! r = Procedures.UpdateUserPassword dbConnectionGetter uid hash passwordConfig

                                return this.Redirect("/") :> IActionResult
                            else
                                viewBag["Error"] <- "Current password is invalid"
                                return this.View(model) :> IActionResult
                        }

                } |> Async.StartAsTask
        else
            task { return this.View(model) }